### PR TITLE
[MIG] [12.0] oauth_provider

### PR DESCRIPTION
**Pull request by hbrunn**
this is the v12 version of https://github.com/OCA/server-tools/tree/9.0/oauth_provider

Note I added a few changes on top of the migration:

- previously, the permissions on clients and some other models were too lax. in v12, only managers have access to clients and scopes, everything else goes via sudo(). That's why a bunch of methods are private now
- implemented client type `backend application`
- implemented http auth method `oauth_provider`

Those involved some necessary refactoring, that's why I rolled that together.

### CODE REVIEW

**Comment by hbrunn** [comment]
tests are fixed in https://github.com/OCA/server-auth/pull/406

**Review by antonioburic — APPROVED** [review]

**Review by astirpe — APPROVED** [review]

**Comment by victor-champonnois** [comment]
@hbrunn Can this be merged ? I'm going to port the code to 14.0 based on this PR

**Comment by hbrunn** [comment]
@victor-champonnois there's no need to wait until this is merged, you can just take the commits from here. Actually, on a higher version the code has a better chance to get merged as probably more people are interested

**Comment by hbrunn** [comment]
CI is fixed in https://github.com/OCA/server-auth/pull/465

File: oauth_provider/models/oauth_provider_token.py
Comment (remytms) [inline oauth_provider/models/oauth_provider_token.py:44]: Shouldn't this method be prefixed by `@api.depends("expires_at")` as the method use this field ?

Comment (hbrunn) [reply oauth_provider/models/oauth_provider_token.py:44]: we need the dependency when we store the field, but we don't

**Comment by remytms** [comment]
Module in 16.0 is [here](https://github.com/OCA/server-auth/pull/795).

**Pull request closed without merging**
